crates/cubecl-runtime/src/memory_management/memory_manage.rs

```diff
@@ -352,6 +337,45 @@ impl<Storage: ComputeStorage> MemoryManagement<Storage> {
     }
 }
 
+/// Trait necessary to implement `find_suitable_memory_pool` on a foreign type
+trait PoolFind<'a> {
+    fn find_suitable_memory_pool(self, size: u64) -> &'a mut DynamicPool;
+}
+
+impl<'a> PoolFind<'a> for &'a mut [DynamicPool] {
+    /// Finds and returns a memory pool big enough for the requested size.
+    ///
+    /// # Panics
+    ///
+    /// Panics if no pool big enough can be found.
+    fn find_suitable_memory_pool(self, size: u64) -> &'a mut DynamicPool {
+        let pool_ind = self
+            // This finds a (not necessarily first) pool that matches the size exactly
+            // or returns error with index where a matching element could be inserted while maintaining sorted order
+            .binary_search_by_key(&size, |p| p.max_alloc_size())
```

Comment:
`size` won't be equal to any of the `max_alloc_sizes`, we should find a bucket where `size >= min_slice_size && size <= max_alloc_size`. We don't need binary search since we only have < 10 buckets, max 20. Having the trait is great though.

Reply:
Sorry, I didn't know about the `size >= min_slice_size` condition. But I also can't find `min_slice_size` defined anywhere.

Is the `Vec<DynamicPools>` sorted by `min_slice_size` too? Because then binary search would still be possible. 

Binary search was what `partition_point()` was doing and this implementation only extends that technique. But with such a small array size I do agree that linear search will not impact the performance and will allow for the `size >= min_slice_size` condition - I'll change the implementation.

Reply:
Hiya - give I wrote this I'll barge in - there is no min_slice_size, that's implicit by the fact that that should have been allocated in the previous pool. Stuff being sorted is therefore a condition anyway - and binary search seems fine? I agree it's a small performance win but... idk, seems clearer & faster. I just didn't quite handle the error condition of the search correctly.

The trait also seems like it could just be a struct? Doubt you'd want multiple possible implementations of this.

Reply:
In case there's only one condition, then the most important consideration for using binary search is that if there are two pools with same `max_alloc_size` then it's not predictable which one will be chosen. The question is if that's an acceptable behaviour.

> The trait also seems like it could just be a struct? Doubt you'd want multiple possible implementations of this.

Struct or newtype would require changing the type of `pools` in `MemoryManagement` and I didn't feel comfortable doing that. So I created the trait, which I admit is a little bit excessive, but necessary for borrow checker.